core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeManager.java

```diff
@@ -359,11 +358,41 @@ public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_
         return getFromSolution(valueRangeDescriptor, cachedWorkingSolution);
     }
 
+    public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor,
+            @Nullable SelectionSorter<Solution_, T> sorter) {
+        if (cachedWorkingSolution == null) {
+            throw new IllegalStateException(
+                    "Impossible state: value range (%s) requested before the working solution is known."
+                            .formatted(valueRangeDescriptor));
+        }
+        return getFromSolution(valueRangeDescriptor, cachedWorkingSolution, sorter);
+    }
+
     @SuppressWarnings("unchecked")
     public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor,
             Solution_ solution) {
-        var valueRange = fromSolution[valueRangeDescriptor.getOrdinal()];
-        if (valueRange == null) { // Avoid computeIfAbsent on the hot path; creates capturing lambda instances.
+        return getFromSolution(valueRangeDescriptor, solution, null);
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor, Solution_ solution,
+            @Nullable SelectionSorter<Solution_, T> sorter) {
+        var item = fromSolution[valueRangeDescriptor.getOrdinal()];
+        var valueRange =
+                item != null ? (CountableValueRange<T>) item.countableValueRange() : null;
+        var valueRangeSorter = item != null ? item.sorter() : null;
+        // The phase initialization logic can call operations like countOnSolution or countOnEntity,
+        // which do not consider sorting and initialize the value range without a sorter.
+        // Therefore, we return the range if there is no sorter or applied sorter is the same as the given sorter
```

Comment:
If we know that there will be a sorter in the future (which we do know), why not just sort this the first time it's called, and avoid all of this complexity?